Provide a consistent interface for specifying nullary constraints

dcr8898 opened the pull request:
Closes #473

Implements the following api for specifying nullary constraints (no additional argument needed, e.g., `:odd`) and unary constraints (additional argument needed, e.g., `gt: 18`).

```ruby
type.constrained(:odd) # new variant
type.constrained(:odd, :even) # new variant
type.constrained(:odd, gt: 200, lt: 300) # new variant
```

The above is covered by new tests and updated yardocs. If this api is acceptable, I will update the docsite accordingly.

## Addenda

The api also accepts the following for backward compatibility. However, I suggest deprecating this behavior as non-intuitive and dependent on a quirk of the former implementation (`#map`). Deprecating will allow for the removal of an otherwise unnecessary application of `Array#flatten` in processing arguments.

```ruby
type.constrained([:odd, :even]) # worked before, continue to support
type.constrained([:odd, :even], gt: 200, lt: 300) # new variant, supported as a side-effect
```
The following also work, although I don't believe they are covered by tests. The last two are no-ops.

```ruby
type.constrained # error, as before
type.constrained([]) # worked before, continue to support
type.constrained({}) # worked before, continue to support
```

Finally, the following also work as undocumented behavior.

```ruby
type.constrained({gteq: 0, lt: 100}) # undocumented, not encouraged
type.constrained({gteq: 0}, lt: 100) # undocumented, not encouraged
```

## Potential Issues

- Dry::Types::Sum.constrained (on line 89) does not double-splat the options Hash used to call `Constrained#constrained`, which means that it relies on the "undocumented" behavior above. Not a big deal, but I would consider adding `**` for clarity or, even better, using argument forwarding (`...`).

- Dry::Types::Builder.constrained does not implement the above api. Although this does not break current tests, this should probably be modified (with tests) to be consistent. Maybe just use argument forwarding here as well?

solnic commented:
```ruby
type.constrained([]) # worked before, continue to support
type.constrained({}) # worked before, continue to support
```

I think we should tweak it to raise an exception in 2.0

> Dry::Types::Sum.constrained (on line 89) does not double-splat the options Hash used to call Constrained#constrained, which means that it relies on the "undocumented" behavior above. Not a big deal, but I would consider adding ** for clarity or, even better, using argument forwarding (...).
> Dry::Types::Builder.constrained does not implement the above api. Although this does not break current tests, this should probably be modified (with tests) to be consistent. Maybe just use argument forwarding here as well?

It should be safe to use `...` there and it would be good to do it already in this PR :)

flash-gordon commented:
> I think we should tweak it to raise an exception in 2.0

Such calls can occur when dry-types is wrapped inside another lib.

dcr8898 commented:
```ruby
type.constrained([]) # worked before, continue to support
type.constrained({}) # worked before, continue to support
```

I don't have an opinion either way on keeping this, but if it's important it should be covered by tests. I don't think it is.

dcr8898 commented:
I have changed Dry::Types::Builder#constrained and Dry::Types::Sum#constrained to use argument forwarding. All tests remain green.

## Issues

- I don't know how to represent argument forwarding in Yardoc, so I didn't touch those.
- Passing an empty Array or Hash to `#constrained` doesn't *really* work (before or after the proposed changes).

### Yardoc

I found this example of argument forwarding in the codebase (for another invocation of `#constrained` no less!), but I wasn't sure if this is the canonical way to represent argument forwarding in Yardoc (as an `[Array]`). It looks suspect to me. :monocle_face: 

```ruby
      # Build a constrained type
      #
      # @param [Array] args see {Dry::Types::Builder#constrained}
      #
      # @return [Default]
      #
      # @api public
      def constrained(...)
        type.constrained(...).default(value)
      end
```
### Empty Array/Hash argument

Passing an empty Array or Hash to `#constrained` does not cause an explicit error either before or after the proposed changes. However, they both actually contain a subtle bug.

Entering `Types::Integer` in IRB produces:

```ruby
=> #<Dry::Types[Constrained<Nominal<Integer> rule=[type?(Integer)]>]>
```

Entering `Types::Integer.constrained([])` or `Types::Integer.constrained({})` should logically produce the same result. However, they actually both produce the following:

```ruby
=> #<Dry::Types[Constrained<Nominal<Integer> rule=[type?(Integer) AND ]>]>
```

That dangling `AND` does not seem to effect the functionality of the Type object, but it *does* break the AST parser. That is, the following works:

```ruby
irb(main):005:0> Types::Integer.to_ast
=> [:constrained, [[:nominal, [Integer, {}]], [:predicate, [:type?, [[:type, Integer], [:input, Undefined]]]]]]
```

But the following causes an error:

```ruby
irb(main):005:0> Types::Integer.constrained([]).to_ast
```

This bug stopped me from adding tests for empty Arrays/Hashes, since the test suite invokes `#to_ast` and then errors out.

## Path forward

Please let me know how you would like for me to proceed. I would be happy to update the Yardocs for the two methods I changed in whatever way you recommend.

As for the bug, do you want to address that in this PR, or separately? It might be as easy as adding a guard clause that checks for `rules.empty?` before calling `#with`, but I don't know the proper value to return from the guard clause.

Guidance is appreciated if you would like for me to investigate further.

flash-gordon commented:
@dcr8898 this looks good overall, I'll now merge it as it is and run my apps against it. If everything's fine, I'll return to the YARD nuances. Thank you for working on this! ❤️

flash-gordon commented:
So far, it works well, I'll look into YARD issues soon
